File: lemur-app/src/main/java/io/github/malczuuu/lemur/app/infra/data/jpa/player/PrimaryPlayerRepository.java
package io.github.malczuuu.lemur.app.infra.data.jpa.player;

import static io.github.malczuuu.lemur.app.common.NumberUtils.safeParseLong;

import io.github.malczuuu.lemur.app.domain.player.Player;
import io.github.malczuuu.lemur.app.domain.player.PlayerRepository;
import java.util.List;
import java.util.Optional;
import org.springframework.stereotype.Component;

@Component
class PrimaryPlayerRepository implements PlayerRepository {

  private final PlayerJpaRepository playerRepository;

  private final PlayerEntityMapper playerMapper = new PlayerEntityMapper();

  PrimaryPlayerRepository(PlayerJpaRepository playerRepository) {
    this.playerRepository = playerRepository;
  }

  @Override
  public List<Player> findAll() {
    return playerRepository.findAll().stream().map(playerMapper::toPlayer).toList();
  }

  @Override
  public Optional<Player> findById(String id) {
    return safeParseLong(id).flatMap(playerRepository::findById).map(playerMapper::toPlayer);
  }

  @Override
  public Optional<Player> lockById(String id) {
    return safeParseLong(id).flatMap(playerRepository::lockById).map(playerMapper::toPlayer);
  }

  @Override
  public Player save(Player player) {
    PlayerEntity entity = playerMapper.toEntity(player);
    entity = playerRepository.save(entity);
    return playerMapper.toPlayer(entity);
  }
}

File: lemur-app/src/main/kotlin/io/github/malczuuu/lemur/app/infra/data/jpa/player/PrimaryPlayerRepository.kt
package io.github.malczuuu.lemur.app.infra.data.jpa.player

import io.github.malczuuu.lemur.app.domain.player.Player
import io.github.malczuuu.lemur.app.domain.player.PlayerRepository
import org.springframework.stereotype.Component

@Component
internal class PrimaryPlayerRepository(
    private val playerRepository: PlayerJpaRepository,
    private val mapper: PlayerEntityMapper,
) : PlayerRepository {

    override fun findAll(): List<Player> = playerRepository.findAll().map(mapper::toPlayer)

    override fun findById(id: String): Player? = id.toLongOrNull()
        ?.let { playerRepository.findById(it).orElse(null) }
        ?.let(mapper::toPlayer)

    override fun lockById(id: String): Player? = id.toLongOrNull()
        ?.let { playerRepository.lockById(it).orElse(null) }
        ?.let(mapper::toPlayer)

    override fun save(player: Player): Player {
        var entity = mapper.toEntity(player)
        entity = playerRepository.save(entity)
        return mapper.toPlayer(entity)
    }
}
